index.js

```diff
@@ -11,50 +11,101 @@ var couch2pg = require('./libs/couch2pg/importer')(
       db, couchdb,
       env.couch2pgDocLimit,
       env.couch2pgChangesLimit),
-    xmlforms = require('./libs/xmlforms/extractor')(db);
-
-var batchLoop = function() {
-  var completeBatch = function(lastSeq) {
-    return couch2pg.storeSeq(lastSeq)
-      .then(function() {
-        log.info('Batch completed up to ' + lastSeq);
-      });
-  };
-
-  log.info('Beginning couch2pg and xmlforms batch');
-
-  return couch2pg.importBatch()
-  .then(function(changes) {
-    var allDocs = changes.deleted.concat(changes.edited);
-    log.debug('There are ' + allDocs.length + ' changes');
-
-    if (allDocs.length > 0) {
-      return xmlforms.extract(allDocs)
-      .then(function() {
-        return completeBatch(changes.lastSeq);
-      })
-      .then(batchLoop);
+    xmlforms = require('./libs/xmlforms/updater')(db);
+
+var fallback = 0;
+var sleepMs = function(errored) {
+  if (errored) {
+    var fallbackMs = fallback * 1000 * 60;
+    if (fallbackMs < env.sleepMs) {
+      fallback++;
+      return fallbackMs;
+    } else {
+      return env.sleepMs;
+    }
+  } else {
+    fallback = 0;
+    return env.sleepMs;
+  }
+};
+
+var migrateCouch2pg = function() {
+  return couch2pgMigrator(env.postgresqlUrl)();
+};
+var migrateXmlforms = function() {
+  return xmlformsMigrator(env.postgresqlUrl)();
+};
+
+var delayLoop = function(loopFn, errored) {
+  return new Promise(function(resolve) {
+    var ms = sleepMs(errored);
+    log.info('Run '+(errored ? 'errored' : 'complete') + '. Next run at ' + new Date(new Date().getTime() + ms));
+    if (ms === 0) {
+      resolve();
     } else {
-      return completeBatch(changes.lastSeq);
+      setInterval(resolve, ms);
     }
   });
 };
 
 var run = function() {
-  log.info('Beginning couch2pg and xmlforms run');
+  log.info('Beginning couch2pg and xmlforms run at ' + new Date());
 
-  return batchLoop()
+  return couch2pg.importAll()
   .catch(function(err) {
+    log.error('Couch2PG import failed');
     log.error(err.stack);
   })
-  .then(function() {
-    log.info('Run complete. Next run at ' + new Date(new Date().getTime() + env.sleepMs));
-    setInterval(run, env.sleepMs);
-  });
+  .then(function(results) {
+    if ((results.deleted.length + results.edited.length) > 0) {
+      return xmlforms.update();
+    }
+  })
+  .then(
+    function() {
+      return delayLoop(run);
+    },
+    function(err) {
+      log.error('XMLForms support failed');
+      log.error(err.stack);
+      return delayLoop(run, true);
+    })
```

Comment:
Usually we use the `.catch` form, rather than passing in two functions to the `then`. Any reason you chose this form this time?